pinot-core/src/main/java/org/apache/pinot/core/util/trace/ContinuousJfrStarter.java

```
@@ -0,0 +1,164 @@
+  ///
+  /// This is only used if toDisk is true.
+  /// The value is a duration string, as defined by the Duration class.
+  /// The default value is 1 day (P1D).
+  ///
+  /// @see #MAX_SIZE
+  public static final String MAX_AGE = "maxAge";
+  public static final String DEFAULT_MAX_AGE = "P1D";
+
+  /// A flag to track whether the JFR recording has been started.
+  /// This is specially useful for testing and quickstarts, where servers, brokers and other components are executed
+  /// in the same JVM.
+  private static boolean _started = false;
+
+  private ContinuousJfrStarter() {
+  }
+
+  public synchronized static void init(PinotConfiguration config) {
+    PinotConfiguration subset = config.subset(CommonConstants.JFR);
+
+    if (!subset.getProperty(ENABLED, DEFAULT_ENABLED)) {
+      return;
+    }
+    if (_started) {
+      return;
+    }
+
+    Path directory = Path.of(subset.getProperty(DIRECTORY, Paths.get(".").toString()));
+
+    Recording recording;
+    String jfrConfName = subset.getProperty(CONFIGURATION, DEFAULT_CONFIGURATION);
+    try {
+      Configuration configuration = Configuration.getConfiguration(jfrConfName);
+      recording = new Recording(configuration);
+    } catch (ParseException e) {
+      throw new RuntimeException("Failed to parse JFR configuration '" + jfrConfName + "'", e);
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to read JFR configuration '" + jfrConfName + "'", e);
+    }
+    boolean dumpOnExit = subset.getProperty(DUMP_ON_EXIT, DEFAULT_DUMP_ON_EXIT);
+    recording.setDumpOnExit(dumpOnExit);
+    if (dumpOnExit) {
+      try {
+        String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss"));
+        String filename = "recording-" + timestamp + ".jfr";
+        Path recordingPath = directory.resolve(filename);
+        boolean newFile = recordingPath.toFile().createNewFile();
```

Why does this file need to be created manually beforehand?

---

And wouldn't it be confusing to use the start time of the process in the file name?

---

> Why does this file need to be created manually beforehand?

Doesn't need to be created. I'm touching it to be sure we have permissions (to fail fast in case we don't). I guess I can check permissions.

> And wouldn't it be confusing to use the start time of the process in the file name?

We need a unique name (so different executions store their own files). Including a timestamp is better than a random number. IMHO, the ideal name would include the instant when the dump was created. But that cannot be done, so the time when the process starts seems to be the best possible solution.

Remember, this is the file name that will be used to create a dump on an ordered JVM exit.